plutus-contract/src/Language/Plutus/Contract/Effects/WatchAddress.hs

```diff
@@ -31,32 +36,98 @@ import           Language.Plutus.Contract.Util              (loopM)
 type AddressSymbol = "address"
 
 type HasWatchAddress s =
-    ( HasType AddressSymbol (Address, Tx) (Input s)
-    , HasType AddressSymbol Address (Output s)
+    ( HasType AddressSymbol AddressChangeResponse (Input s)
+    , HasType AddressSymbol AddressChangeRequest (Output s)
     , ContractRow s)
 
-type WatchAddress = AddressSymbol .== ((Address, Tx), Address)
+type WatchAddress = AddressSymbol .== (AddressChangeResponse, AddressChangeRequest)
+
+-- | Information about transactions that spend or produce an output at
+--   an address. The transactions are returned in the order they appear
+--   in the ledger (oldest first).
+data AddressChangeResponse =
+    AddressChangeResponse
+        { acrAddress :: Address
+        , acrSlot    :: Slot
+        , acrTx      :: Tx
+        }
+        deriving stock (Eq, Generic, Show)
+        deriving anyclass (ToJSON, FromJSON)
+
+instance Pretty AddressChangeResponse where
+    pretty AddressChangeResponse{acrAddress, acrTx, acrSlot} =
+        hang 2 $ vsep
+            [ "Address:" <+> pretty acrAddress
+            , "Slot:" <+> pretty acrSlot
+            , "TxId:" <+> pretty (txId acrTx)
+            ]
+
+data AddressChangeRequest =
+    AddressChangeRequest
+        { acreqSlot    :: Slot
+        , acreqAddress :: Address
+        , acreqTxId    :: Maybe TxId
+        }
+        deriving stock (Eq, Generic, Show, Ord)
+        deriving anyclass (ToJSON, FromJSON)
+
+instance Pretty AddressChangeRequest where
+    pretty AddressChangeRequest{acreqSlot, acreqAddress, acreqTxId} =
+        hang 2 $ vsep
+            [ "Slot:" <+> pretty acreqSlot
+            , "Address:" <+> pretty acreqAddress
+            , "TxId:" <+> pretty acreqTxId
+            ]
+
+{-| Wait for the next transaction that changes an address. The meaning of
+    "next transaction" for an 'AddressChangeRequest' depends on the value
+    of 'acreqTxId'.
+
+    1. If 'acreqTxId' is 'Just txid' then @b@ is the next transaction if
+    @b@ spends or produces an output at the given address and @b@ is the first
+    transaction to be added to the ledger in or after 'acreqSlot' and
+    after 'txid'.
+    2. If 'acreqTxId' is 'Nothing' then @b@ is the next transaction if
+    b spends or produces an output at the given address and b is the first
+    transaction to be added to the ledger in or after 'acreqSlot'.
+
+    Some implications of this definition are:
+
+    * If the transaction 'txid' is never added to the ledger then (1) will
+    never return
+    * If the slot 'acreqSlot' contains more than one transaction at the address
+    then only (1) will ever return transactions other than the first one.
+
+    To get all transactions that modify the address in a slot, it is prudent to
+    call 'addressChangeRequest' repeatedly, using the ID of the last
+    transaction that was returned
```

Comment:
You wouldn't need this with my suggestion (and then perhaps you don't need the version with the txid at all?).

Reply:
Yes it's much simper that way. Thanks.